### PR TITLE
[MemoryBanking] Support memory banking for`GetGlobalOp`

### PR DESCRIPTION
**Pull request by jiahanxie353**
This patch supports memory banking for multi-dimensional `GetGlobalOp`s and restructure the data in their corresponding `GlobalOp`s by the banking scheme (such as the banking factor and the dimension).

### CODE REVIEW

File: test/Transforms/memory_banking_multi_dim.mlir [hunk 1]
```diff
@@ -1,4 +1,5 @@
 // RUN: circt-opt %s -split-input-file -memory-banking="banking-factor=2 dimension=1" | FileCheck %s --check-prefix RANK2-BANKDIM1
+// RUN: circt-opt %s -split-input-file -memory-banking="banking-factor=2" | FileCheck %s --check-prefix GETGLOBAL
```
Comment (cgyurgyik) [inline test/Transforms/memory_banking_multi_dim.mlir:2]: You may have handled this in an earlier PR, but what about corner cases, e.g., `banking-factor=0`, `dim_size % banking-factor != 0`, etc.?

File: lib/Transforms/MemoryBanking.cpp [hunk 2]
```diff
@@ -125,6 +171,76 @@ SmallVector<Value, 4> createBanks(Value originalMem, uint64_t bankingFactor,
             banks.push_back(bankAllocaOp);
           }
         })
+        .Case<memref::GetGlobalOp>([&](memref::GetGlobalOp getGlobalOp) {
```
Comment (cgyurgyik) [inline lib/Transforms/MemoryBanking.cpp:174]: How difficult would it be to place all this code in a separate function and just call it here? It is quite large, but unsure if that would require too many arguments.

**Review by cgyurgyik — COMMENTED** [review]
Nice job! Mostly style nits and a few smaller comments.

File: test/Transforms/memory_banking_multi_dim.mlir (hunk 1 above)
Comment (jiahanxie353) [reply test/Transforms/memory_banking_multi_dim.mlir:2]: Yes, in these cases, it will early signal failure: https://github.com/llvm/circt/blob/4a062a2a1cda5052f61c700aeb204ac4ba7d2f55/lib/Transforms/MemoryBanking.cpp#L381
or raise an assertion error: https://github.com/llvm/circt/blob/4a062a2a1cda5052f61c700aeb204ac4ba7d2f55/lib/Transforms/MemoryBanking.cpp#L80

File: lib/Transforms/MemoryBanking.cpp [hunk 3]
```diff
@@ -88,6 +89,51 @@ MemRefType computeBankedMemRefType(MemRefType originalType,
   return newMemRefType;
 }
 
+// Updates n-dimensional index array `ndIndex` in-place by decoding the flat
+// index `linIndex` given the shape of the array. Assume row-major order.
+void decodeIndex(int64_t linIndex, ArrayRef<int64_t> shape,
```
[the remark this answers is not in the document]
Comment (jiahanxie353) [reply lib/Transforms/MemoryBanking.cpp:94]: sounds good!

File: lib/Transforms/MemoryBanking.cpp [hunk 4]
```diff
@@ -88,6 +89,51 @@ MemRefType computeBankedMemRefType(MemRefType originalType,
...
+
+// Performs multi-dimensional slicing on `allAttrs` by extracting all elements
+// whose coordinates range from `bankCnt`*`bankingDimension` to
+// (`bankCnt`+1)*`bankingDimension` from `bankingDimension`'s dimension, leaving
+// other dimensions alone.
+SmallVector<SmallVector<Attribute>> sliceSubBlock(ArrayRef<Attribute> allAttrs,
+                                                  ArrayRef<int64_t> memShape,
+                                                  unsigned bankingDimension,
+                                                  unsigned bankingFactor) {
+  auto totalSize = [](auto &&shape) {
```
[the remark this answers is not in the document]
Comment (jiahanxie353) [reply lib/Transforms/MemoryBanking.cpp:114]: indeed!

**Comment by jiahanxie353** [comment]
Thanks @cgyurgyik for the review and suggestions! I've finished modifying the code based on the request. And it's ready for another round of review!

File: lib/Transforms/MemoryBanking.cpp [hunk 5]
```diff
@@ -88,6 +90,121 @@ MemRefType computeBankedMemRefType(MemRefType originalType,
...
+  builder.setInsertionPointAfter(globalOp);
+  OpBuilder::InsertPoint globalOpsInsertPt = builder.saveInsertionPoint();
+  builder.setInsertionPointAfter(getGlobalOp);
+  OpBuilder::InsertPoint getGlobalOpsInsertPt = builder.saveInsertionPoint();
+
+  for (size_t bankCnt = 0; bankCnt < bankingFactor; ++bankCnt) {
+    // Prepare relevant information to create a new GlobalOp
+    auto newMemRefTy = MemRefType::get(newShape, globalOpTy.getElementType());
+    auto newTypeAttr = TypeAttr::get(newMemRefTy);
+    std::string newNameStr =
```
Comment (cgyurgyik) [inline lib/Transforms/MemoryBanking.cpp:175]: Nit: It is clear from the type already (std::string), IMO we don't need the `Str` suffix here.
```suggestion
    std::string newName =
```

File: test/Transforms/memory_banking_multi_dim.mlir (hunk 1 above)
Comment (cgyurgyik) [reply test/Transforms/memory_banking_multi_dim.mlir:2]: Awesome!

**Review by cgyurgyik — APPROVED** [review]
LGTM, thanks!

File: lib/Transforms/MemoryBanking.cpp (hunk 5 above)
Comment (jiahanxie353) [reply lib/Transforms/MemoryBanking.cpp]: sounds reasonable!

Plan to merge the patch once the CI has passed. Thanks again for the review :-)
